lib/tokenizer.ts

```diff
@@ -295,5 +313,49 @@ export const tokenizer = (regexpStr: string): Root => {
     );
   }
 
+  updateReferences(referenceQueue, groupCount);
+
   return start;
 };
+
+/**
+ * This is a side effecting function that changes references to chars
+ * if there are not enough capturing groups to reference
+ * See: https://github.com/fent/ret.js/pull/39#issuecomment-1006475703
+ * See: https://github.com/fent/ret.js/issues/38
+ * @param {(Reference | Char)[]} referenceQueue
+ * @param {number} groupCount
+ * @returns {void}
+ */
+function updateReferences(referenceQueue: ReferenceQueue, groupCount: number) {
+  // Note: We go through the queue in reverse order so
+  // that index we use is correct even if we have to add
+  // multiple tokens to one stack
+  for (const elem of referenceQueue.reverse()) {
+    if (groupCount < elem.reference.value) {
+      // If there is nothing to reference then turn this into a char token
+      elem.reference.type = types.CHAR;
+
+      const valueString = elem.reference.value.toString();
+      // If the number is not octal then we need to create multiple tokens
+      // https://github.com/fent/ret.js/pull/39#issuecomment-1008229226
+      if (!/^[0-7]+$/.test(valueString)) {
+        const val = parseInt(valueString[0], 10);
+        elem.reference.value = val > 7 ? valueString.charCodeAt(0) : val;
+
+        if (valueString.length > 1) {
+          const tail = elem.stack.splice(elem.index + 1);
+
+          // eslint-disable-next-line max-depth
```

Comment:
i think we can remove this rule instead. too inconvenient sometimes. what do you think?